test/jdk/com/sun/net/httpserver/bugs/B6968351.java

```diff
@@ -0,0 +1,95 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 6968351
+ * @summary  tcp no delay not required for small payloads
+ * @library /test/lib
+ * @run main B6968351
+ * @run main/othervm -Dsun.net.httpserver.nodelay=false -Djdk.httpclient.HttpClient.log=all -Djava.net.preferIPv6Addresses=true -Djavax.net.debug=all B6968351
+ */
+
+import com.sun.net.httpserver.*;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.*;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.logging.*;
+
+public class B6968351 {
+
+    public static final Logger LOGGER = Logger.getLogger("sun.net.www.protocol.http");
+
+    public static void main (String[] args) throws Exception {
+
+        java.util.logging.Handler outHandler = new StreamHandler(System.out,
+                new SimpleFormatter());
+        outHandler.setLevel(Level.FINEST);
+        LOGGER.setLevel(Level.FINEST);
+        LOGGER.addHandler(outHandler);
+
+        InetAddress loopback = InetAddress.getLoopbackAddress();
+        InetSocketAddress addr = new InetSocketAddress (loopback, 0);
+        HttpServer server = HttpServer.create (addr, 0);
+        HttpHandler handler = new Handler();
+        HttpContext ctx = server.createContext ("/test", handler);
+        ExecutorService executor = Executors.newCachedThreadPool();
+        server.setExecutor (executor);
+        server.start ();
+
+        HttpClient client = HttpClient.newBuilder().build();
+
+        long start = System.currentTimeMillis();
+        for(int i=0;i<1000;i++) {
+            var response = client.send(HttpRequest.newBuilder(new URI("http://localhost:"+server.getAddress().getPort()+"/test")).build(), HttpResponse.BodyHandlers.ofString());
+            if(!response.body().equals("hello")) throw new IllegalStateException("incorrect body");
+        }
+        long time = System.currentTimeMillis()-start;
+        System.out.println("time "+time);
+        if(time>5000) throw new IllegalStateException("took too long");
+        server.stop(0);
+        executor.shutdown();
+    }
+
+    static class Handler implements HttpHandler {
+        public void handle (HttpExchange t)
+                throws IOException
+        {
+            InputStream is = t.getRequestBody();
+            Headers map = t.getRequestHeaders();
+            Headers rmap = t.getResponseHeaders();
+            while (is.read () != -1) ;
+            is.close();
+            rmap.add("content-type","text/plain");
+            t.sendResponseHeaders(200,5);
```

Comment:
if I read the code correctly, there might be a similar delay when sending the last chunk of a chunked response. Would you like to fix it here as well? If not, we can file another ticket for it.

In order to send a chunked response, change the second parameter to `0` here.

Reply:
If you send a chunked response you must close the exchange, or close the output stream, or there is no way to determine the data is done.

Reply:
The problem with closing a chunked stream is that it flushes both before and after the final chunk. I assume it may also cause delays.

The only case when you don't need to close the exchange or the output stream is when your response contains no content.

Reply:
> The problem with closing a chunked stream is that it flushes both before and after the final chunk. I assume it may also cause delays.
> 
> The only case when you don't need to close the exchange or the output stream is when your response contains no content.

I don't understand this in reference to the other tests failing because they are not closing the output stream or ending the exchange. If that is incorrect behavior - because they are sending content - then those tests are invalid.

Still I have a fix for this I believe.

Reply:
ah no, this is not related to the test failures; this is only about the original issue (small sends result in problems with delayed acks)

Reply:
> The problem with closing a chunked stream is that it flushes both before and after the final chunk. I assume it may also cause delays.
> 
> The only case when you don't need to close the exchange or the output stream is when your response contains no content.

Can you point me to the code covering the before and after the final chunk? I would think the before flush should be removed.

Reply:
https://github.com/openjdk/jdk/blob/b704e91241b0f84d866f50a8f2c6af240087cb29/src/jdk.httpserver/share/classes/sun/net/httpserver/ChunkedOutputStream.java#L140-L144

Reply:
I figured out what you are referring to. Let me look at this, I am pretty sure I can remove the first flush.

Reply:
I posted a fix which addresses this. I will modify the test case to test both methods of returning the data.

Reply:
test case has been updated as well to test both fixed and chunked responses.